harvester_e2e_tests/integration/test_upgrade.py

```diff
@@ -1,121 +1,1049 @@
+import os
+import uuid
+from time import sleep, time
+from datetime import datetime, timedelta
+
 import pytest
+import requests
+from paramiko import SSHClient, AutoAddPolicy
+
+DEFAULT_STORAGE_CLS = "harvester-longhorn"
+
+DEFAULT_USER = "ubuntu"
+DEFAULT_PASSWORD = "root"
+
+DEFAULT_NODE_USERNAME = "rancher"
+DEFAULT_NODE_PASSWORD = "root"
+
+DOWNLOAD_ISO_DEFAULT_URL_FMT = "https://releases.rancher.com"
+DOWNLOAD_ISO_PATH_FMT = "{URL}/harvester/{VERSION}/harvester-{VERSION}-amd64.iso"
+DOWNLOAD_ISO_CHECKSUM_PATH_FMT = "{URL}/harvester/{VERSION}/harvester-{VERSION}-amd64.sha512"
+
+NETWORK_VLAN_ID_LABEL = "network.harvesterhci.io/vlan-id"
+UPGRADE_STATE_LABEL = "harvesterhci.io/upgradeState"
+CONTROL_PLANE_LABEL = "node-role.kubernetes.io/control-plane"
+NODE_INTERNAL_IP_ANNOTATION = "rke2.io/internal-ip"
+
+LOGGING_NAMESPACE = "cattle-logging-system"
+HARVESTER_NAMESPACE = "harvester-system"
+LONGHORN_NAMESPACE = "longhorn-system"
+
+expected_harvester_crds = {
+    "addons.harvesterhci.io": False,
+    "blockdevices.harvesterhci.io": False,
+    "keypairs.harvesterhci.io": False,
+    "preferences.harvesterhci.io": False,
+    "settings.harvesterhci.io": False,
+    "supportbundles.harvesterhci.io": False,
+    "upgrades.harvesterhci.io": False,
+    "versions.harvesterhci.io": False,
+    "virtualmachinebackups.harvesterhci.io": False,
+    "virtualmachineimages.harvesterhci.io": False,
+    "virtualmachinerestores.harvesterhci.io": False,
+    "virtualmachinetemplates.harvesterhci.io": False,
+    "virtualmachinetemplateversions.harvesterhci.io": False,
+
+    "clusternetworks.network.harvesterhci.io": False,
+    "linkmonitors.network.harvesterhci.io": False,
+    "nodenetworks.network.harvesterhci.io": False,
+    "vlanconfigs.network.harvesterhci.io": False,
+    "vlanstatuses.network.harvesterhci.io": False,
+
+    "ksmtuneds.node.harvesterhci.io": False,
+
+    "loadbalancers.loadbalancer.harvesterhci.io": False,
+}
+
+pytest_plugins = [
+    "harvester_e2e_tests.fixtures.api_client",
+]
+
+
+def _lookup_image(api_client, name):
+    code, data = api_client.images.get()
+    if code == 200:
+        for image in data['items']:
+            if image['metadata']['name'] == name:
+                return image
+    return None
+
+
+def _create_image(api_client, url, name, wait_timeout=300):
+    image_data = _lookup_image(api_client, name)
+    if image_data is None:
+        code, image_data = api_client.images.create_by_url(name, url)
+        assert code == 201, "failed to create image"
```

Comment:
try to append the data returned after the error message should improve debug ability, for example `assert code == 201, f"Failed to create image: {image_data}"`